catalog/db_branch_scanner.go

```diff
@@ -35,9 +42,37 @@ func NewDBBranchScanner(tx db.Tx, branchID int64, commitID CommitID, opts *DBSca
 		s.opts.BufferSize = DBScannerDefaultBufferSize
 	}
 	s.buf = make([]*DBScannerEntry, 0, s.opts.BufferSize)
+	commitsWhere, _ := getRelevantCommitsCondition(tx, branchID, commitID)
+	s.commitsWhere = commitsWhere
 	return s
 }
 
+func getRelevantCommitsCondition(tx db.Tx, branchID int64, commitID CommitID) (string, error) {
+	var branchMaxCommitID CommitID
+	var commits []string
+	var commitsWhere string
+	if commitID == UncommittedID || commitID == CommittedID {
+		branchMaxCommitID = MaxCommitID
+	} else {
+		branchMaxCommitID = commitID
+	}
+	// commit_id name is changed so that sorting will be performed on the numeric value, not the string value (where "10" is less than "2")
+	sql := "SELECT commit_id::text as str_commit_id FROM catalog_commits WHERE branch_id = $1 AND commit_id <= $2 ORDER BY commit_id"
+	err := tx.Select(&commits, sql, branchID, branchMaxCommitID)
+	if err != nil {
+		panic(err)
+	}
+	if commitID == UncommittedID {
+		commits = append(commits, strconv.FormatInt(int64(MaxCommitID), 10))
+	}
+	if len(commits) < MaxCommitsInFilter {
+		commitsWhere = "min_commit in (" + strings.Join(commits, `,`) + ")"
```

Comment:
Does this work when no commits are found?

Reply:
yes